### ISSUE TITLE
PgBackrest 2.31 memory leak?

### ISSUE THREAD

**Issue by kolbasky** [pgbackrest/pgbackrest#1319]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:    2.31

2. PostgreSQL version:    12.5

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
PostgreSQL Patroni cluster 3 nodes - CentOS Linux release 7.9.2009 (4 cores, 16 GB ram)
Repository host - CentOS Linux release 7.7.1908 (2 cores, 6GB ram)

4. Did you install pgBackRest from source or from a package?
package

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
**Repo host**
```
[global]
backup-standby=y
process-max=3
repo1-path=/var/lib/pgbackrest
repo1-retention-full=10
start-fast=y

[zabbixpg]
pg1-host=10.x.x.x
pg1-path=/var/lib/pgsql/12/data
pg2-host=10.x.x.x
pg2-path=/var/lib/pgsql/12/data
pg3-host=10.x.x.x
pg3-path=/var/lib/pgsql/12/data
db1-socket-path=/var/run/postgresql
backup-standby=y
repo1-retention-full=2
process-max=2
```

`DB host`
```
[zabbixpg]
pg1-path=/var/lib/pgsql/12/data
db1-socket-path=/var/run/postgresql

[global]
log-level-file=detail
repo1-host=10.x.x.x
archive-async=y
archive-push-queue-max=40GB
spool-path=/var/lib/pgbackrest/spool
compress-type=lz4

[global:archive-get]
process-max=3

[global:archive-push]
process-max=3
compress-level=3
```

    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
```
archive_command=pgbackrest --stanza=zabbixpg archive-push %p
archive_mode=on
listen_addresses=127.0.0.1,10.x.x.x
max_wal_senders=15
wal_level=replica
port=5432
```

    - errors in the postgresql log file before or during the time you experienced the issue
nothing interesting
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)
On repo-host
-------------------PROCESS START-------------------
2021-03-01 14:17:50.157 P00   INFO: backup command begin 2.31: --archive-copy --backup-standby --exec-id=48628-922fed87 --log-level-console=info --pg1-host=10.x.x.x --pg2-host=10.x.x.x --pg3-host=10.x.x.x --pg1-path=/var/lib/pgsql/12/data --pg2-path=/var/lib/pgsql/12/data --pg3-path=/var/lib/pgsql/12/data --pg1-socket-path=/var/run/postgresql --process-max=2 --repo1-path=/var/lib/pgbackrest --repo1-retention-full=2 --stanza=zabbixpg --start-fast --type=full
2021-03-01 14:18:02.134 P00   INFO: execute non-exclusive pg_start_backup(): backup begins after the requested immediate checkpoint completes
2021-03-01 14:18:04.454 P00   INFO: backup start archive = 0000000400004EDC0000009E, lsn = 4EDC/9E3151C8
2021-03-01 14:18:04.454 P00   INFO: wait for replay on the standby to reach 4EDC/9E3151C8
2021-03-01 14:18:06.669 P00   INFO: replay on the standby reached 4EDC/9E3151C8
2021-03-01 14:18:10.816 P00   WARN: resumable backup 20210301-020011F of same type exists -- remove invalid files and resume
2021-03-01 14:18:47.487 P01   INFO: backup file 10.x.x.x:/var/lib/pgsql/12/data/log/postgresql-Mon.log (69.2MB, 0%) checksum 6771b76fcd546443eb4186817997fb4921745d4e
...............................................................
2021-03-01 21:24:55.512 P03   INFO: backup file 10.x.x.x:/var/lib/pgsql/12/data/base/14184/14022 (0B, 100%)
2021-03-01 21:24:55.552 P00   INFO: full backup size = 132.8GB
2021-03-01 21:24:55.552 P00   INFO: execute non-exclusive pg_stop_backup() and wait for all WAL segments to archive
2021-03-01 21:24:55.757 P00   INFO: backup stop archive = 0000000400004F0F00000070, lsn = 4F0F/70E71B98
2021-03-01 21:24:55.812 P00   INFO: check archive for segment(s) 0000000400004F090000001B:0000000400004F0F00000070

7. Describe the issue:

We have a zabbix database in patroni cluster. The db is about 140GB.
First we had default compression settings for pgbackrest, but it couldn't keep up with wal generation tempo, so eventualy we got  errors in logs like "wal file dropped because exceeded push queue".  But full backups were taken with no problems.
After that, we switched to compression_type=lz4 and now wal archiving works like a charm. 
But unfortunately we can't take full backup as it always gets stuck on "check archive for segment(s)".
It hangs like that until  pgbackrest process eats out all the memory and get killed by OOM.

I take full backups with the following command:
pgbackrest --stanza=zabbixpg backup --log-level-console=info --type=full --archive-copy

**Comment by kolbasky** [comment]
Just realized, i have compress-type=lz4 specified in config on DB host, but i don't have that option on repo-host. 
Added to [global]
compress-type=lz4

and now full backup has been successfully taken.

seems like some kind of incompatibility in compression types. maybe this should be pointed out in documentation? or there should be some kind of check - when wal archive compression is not equal to backup compression.

**Comment by dwsteele** [comment]
Well, looks a lot more like a memory deluge than a leak to me. :)

We're definitely leaking memory in this loop and it gets a *lot* worse if the files need to be re-compressed in a different format. The manifest format does not allow files in the backup to have different compression types, so they need to be converted. The best answer is to use the same compression for backup and archive, as you have done.

We'll fix the leak and maybe it would also make sense to add a log message if the file needs to be re-compressed since that will take a long time.

**Comment by dwsteele** [comment]
Also, how would you like your name to appear in the release notes?

**Comment by kolbasky** [comment]
Thanks for quick reply!

I'd prefer Efremov Egor over kolbasky :-)

**Comment by dwsteele** [comment]
Fixed at 778adbf19f71ceb9.

**Comment by dwsteele** [comment]
Opened PRs #1327 and #1328 for additional documentation and a log message.
